### PR TITLE
Add screenshot for single screen

### LINKED ISSUE #30: Taking screenshot of one display.
How do I make shotgun screenshot a single display?

### PR DESCRIPTION
**Pull request by ugnmura**
Resolves #30

### CODE REVIEW

**Comment by ugnmura** [comment]
Sorry for the diffs, my formatter did that automatically.

**Review by Lonami — APPROVED** [review]
Assuming it builds changes look fine, although perhaps we should separate the format and changes into two commits.

File: src/main.rs
Comment (pickfire) [inline src/main.rs]: I don't think this change is necessary.

**Comment by pickfire** [comment]
It doesn't seem to update the docs, other people would have to look at the code to know.

**Comment by 9ary** [comment]
> It doesn't seem to update the docs, other people would have to look at the code to know.

~~They did add help text for the new flag, but it should indeed be copied over to the README.~~
I'm blind. This doesn't seem to add a new flag at all?

**Comment by bb010g** [comment]
> Sorry for the diffs, my formatter did that automatically.

Please split the reformatting out into a separate, dedicated commit.

File: src/main.rs
Comment (9ary) [inline src/main.rs]: Ok, I see what you did here. This means that `-g` without specifying a region would capture the screen where the cursor is. This isn't good, please add a separate flag instead, and document it properly.

**Review by bb010g — CHANGES_REQUESTED** [review]
I agree with @9ary. Please do not introduce options that optionally take values to this CLI. The functionality looks fine, if implemented differently.

**Comment by ugnmura** [comment]
Hi, I appreciate the reviews, my suggestion would be to revert the commit once and
1. separate the commits from functionality and formatting
2. introduce the new flag
3. add more documentation

Unfortunately I do not have much time, and therefore I cannot implement it right away.

File: src/main.rs
Comment (9ary) [inline src/main.rs]: I don't think splitting this line in two or even specifying the type is strictly necessary here. Is it?

File: src/main.rs
Comment (9ary) [inline src/main.rs]: This introduces an additional call to `get_screen_rects` which is a little inefficient. It should be cached, or better yet lazily initialized so it only gets called when needed. [`std::Lazy`](https://doc.rust-lang.org/std/lazy/struct.Lazy.html) is currently experimental, so we'll need to find a crate for this.
Since `get_screen_rects` returns an iterator, you'll need to collect it into a vector, and then make the masking step reuse that. Maybe removing the iterator altogether would be a good idea. I believe I originally wrote it this way to avoid superfluous allocations, but that's already out of the window, so there's no point in keeping it.

We also already cache the result of `get_default_root` during initialization. You should definitely use that.

File: src/main.rs
Comment (9ary) [inline src/main.rs]: Alright so under normal circumstances this shouldn't fail. That said, maybe we should only bother to mutate `sel` if something is found? Honestly not sure what the best strategy is here. This seems sound enough all things considered.

File: src/util.rs
Comment (9ary) [inline src/util.rs]: I think `Point` would be a better name for this struct.

File: src/xwrap.rs
Comment (9ary) [inline src/xwrap.rs]: This function returns an `Option`, but it doesn't appear to handle the failure case. If `XQueryPointer` is fallible, then it should return `None` on failure. Otherwise, get rid of the `Option`.

File: src/main.rs
Comment (9ary) [inline src/main.rs]: I think something along the lines of `single-screen` or `current-screen` might be a better long name for this option. `--screen` isn't very descriptive when reading a script.
Also, "current screen" is a bit too vague. Make it clear that this is determined by cursor location.

**Review by 9ary — CHANGES_REQUESTED** [review]
Looking pretty good overall.

Another concern I have is you don't seem to handle passing both `-g` and `-s` gracefully. Currently, `-s` will override `-g` unconditionally. I would prefer to either switch that around, or take the intersection.

File: src/xwrap.rs
Comment (9ary) [reply src/xwrap.rs]: Also I see another call to `get_default_root`. This method should probably be implemented for `xlib::Window` instead, or take the root window as another parameter.

**Comment by bb010g** [comment]
> Another concern I have is you don't seem to handle passing both `-g` and `-s` gracefully. Currently, `-s` will override `-g` unconditionally. I would prefer to either switch that around, or take the intersection.

On this note, I would highly recommend at least reading the [Command Line Arguments section of Fuchsia's Command-line Tools Rubric](https://fuchsia.dev/fuchsia-src/development/api/cli#command_line_arguments). Here's a relevant section from it:

> ##### Mutually Exclusive Options
>
> Some options don't make sense with other options. We call the options mutually exclusive.
>
> Passing mutually exclusive options is considered a user error. When this occurs the tool will do one of the following:
>
> - Write an error message explaining the issue and exit with a non-zero result code; doing no work (i.e. there was no data changed as a result of the call). This is the expected handling, so no further documentation or notes are required.
> - Prioritize one option over another. E.g. "`passing -z will override -y`". In this case the handling will be documented in the `--help` output.
> - Other handling is possible (first takes precedence or last takes precedence or something else) though this is discouraged. In this case the handling will be documented in the Description, Options, ***and*** Notes; though "`See Notes`" may be used in Description and Options with the full write-up in `Notes`.

**Comment by 9ary** [comment]
Erroring out is also an acceptable solution yes. Either way. the conflict behavior should be documented.

File: src/main.rs
Comment (ugnmura) [reply src/main.rs]: As for this part, I think it should be fine if we cache it for now using a vector, I think we can still change that later on.

Comment (9ary) [reply src/main.rs]: Yes, do that, and adapt the masking step to reuse the same vector. Lazy evaluation might not even be necessary considering that masking always happens so forget about that.

File: src/xwrap.rs
Comment (9ary) [inline src/xwrap.rs]: So according to the documentation, XQueryPointer *is* fallible. https://tronche.com/gui/x/xlib/window-information/XQueryPointer.html

The first failure mode is a BadWindow error if you pass in an invalid window. For our use that would be a programming error so you can just make that a panic. (actually xlib will very likely do that itself so maybe it's not necessary, try passing it some random integer and see what happens)

It looks like the other failure case is related to having multiple X screens (not the same as multiple monitors). That is a very uncommon configuration that we shouldn't bother supporting. Regardless, you should re-introduce the Option and the error message for when XQueryPointer returns false.

Lastly, the root_win argument should just be called window, and it shouldn't be mutable. This is a big C-ism that we don't have to deal with in Rust. If you want to use the values XQueryPointer puts in there, you should put them in the return struct. It's unnecessary in this case though, so you should just throw root_return and child_return away.

I do wonder, what happens if you pass 0 as the window argument? If it works, then we should still get the cursor position relative to the root window and you can get rid of the window argument. Otherwise, win_x and win_y are the values you should be returning, as they are relative to the window you're passing in.

**Review by 9ary — CHANGES_REQUESTED** [review]
I think -s should also conflict with -i. -g + -i makes sense, and we just take the intersection there, but that wouldn't make sense with -s.

File: src/xwrap.rs
Comment (ugnmura) [reply src/xwrap.rs]: Ok, I'll make sure that I get that done soon, unfortunately this is the first project I worked on with rust so I'll have to take a closer look to the error handing part.

File: src/main.rs
Comment (pickfire) [inline src/main.rs]:
```suggestion
        Some(r) => r.cloned().collect(),
```

File: src/main.rs
Comment (9ary) [inline src/main.rs:149]: This means that shotgun will always bail out if screens can't be enumerated, even when it's not strictly required. XRandR should be available on pretty much anything these days, do we care enough to support the failure case?

**Comment by ugnmura** [comment]
Alright I feel like I fixed all the stuff mentioned, did I miss anything?

**Comment by 9ary** [comment]
I'm still not sure the usage of XQueryPointer is correct, but I currently don't have the energy to continue the back and forth on this pull request (for reasons unrelated). If it works in its current state then I think it's okay to merge this as it is and we can fix that wrapper later.

I'll leave it up to someone who can actually test this to merge the PR and prepare a new release.

**Review by 9ary — APPROVED** [review]

**Review by Lonami — CHANGES_REQUESTED** [review]
There's still some formatting changes but I don't think that's worth fighting over. Implementation seems fine if it gets the job done, Most if not all concerns have also been fixed so with @9ary's approval I'm going to go ahead and merge it. Thanks!

**Review by Lonami — APPROVED** [review]
(I meant to approve, not request changes.)

**Pull request merged**
